proposals/inline-classes.md

```diff
@@ -0,0 +1,296 @@
+    override fun compareTo(other: Name): Int = s.compareTo(other.s)
+    
+    fun greet() {
+        println("Hello, $s")
+    }
+}    
+
+fun greet() {
+    val name = Name("Kotlin") // there is no actual instantiation of class `Name`
+    name.greet() // method `greet` is called as a static method
+}
+```
+
+## Current limitations
+
+Currently, inline classes must satisfy the following requirements:
+
+- Inline class must have a public primary constructor with a single value parameter
+- Inline class must have a single read-only (`val`) property as an underlying value, which is defined in primary constructor
+- Underlying value cannot be of the same type that is containing inline class
+- Inline class with undefined (recursively defined) generics, e.g. generics with an upper bound equal to the class, is prohibited
+    ```kotlin
+    inline class A<T : A<T>>(val x: T) // error
+    ```
+- Inline class cannot have `init` block
+- Inline class must be final
+- Inline class can implement only interfaces
+- Inline class cannot have backing fields
+    - Hence, it follows that inline class can have only simple computable properties (no lateinit/delegated properties)
+- Inline class cannot have inner classes
+- Inline class must be a toplevel class 
+
+Sidenotes:
+
+- Let's elaborate requirement to have public primary constructor and restriction of `init` blocks.
+For example, we want to have an inline class for some bounded value:
+    ```kotlin
+    inline class Positive(val value: Int) {
+        init { 
+            assert(value > 0) "Value isn't positive: $value" 
+        }
+    }
+  
+    fun foo(p: Positive) {}
+    ```
+    
+    Because of inlining, method `foo` have type `int` from Java POV, so we can pass to method `foo` everything we want and `init` 
+    block will not be executed. Since we cannot control behaviour of `init` block execution, we restrict it for inline classes.
+    
+    Unfortunately, it's not enough, because `init` blocks can be emulated via factory methods:
+    ```kotlin
+    inline class Positive private constructor(val value: Int) {
+        companion object {
+            fun create(x: Int) {
+                assert(x > 0) "Value isn't positive: x"
+                return Positive(x)  
+            }  
+        }
+    }
+  
+    fun foo(p: Positive) {}
+    ```
+    
+    Again, method `foo` have type `int` from Java POV, so we can indirectly create values of type `Positive` 
+    even with the presence of private constructor.
+    
+    To make behaviour more predictable and consistent with Java, we demand public primary constructor and restrict `init` blocks.
+
+### Other restrictions
+
+The following restrictions are related to the usages of inline classes:
+
+- Referential equality (`===`) is prohibited for inline classes
+- vararg of inline class type is prohibited
+```
+inline class Foo(val s: String)
+
+fun test(vararg foos: Foo) { ... } // should be an error  
+```
+
+## Java interoperability  
```

Comment:
I guess we need JS and Native interoperability sections as well :)

Reply:
At some point we'll have it :) 
For now, I'm going to put more information about common and expect/actual inline classes